core/src/main/java/org/apache/iceberg/actions/BaseCommitService.java

```diff
@@ -0,0 +1,224 @@
+    LOG.info("Closing commit service for {} waiting for all commits to finish", table);
+    committerService.shutdown();
+
+    boolean timeout = false;
+    try {
+      // All rewrites have completed and all new files have been created, we are now waiting for
+      // the commit
+      // pool to finish doing its commits to Iceberg State. In the case of partial progress this
+      // should
+      // have been occurring simultaneously with rewrites, if not there should be only a single
+      // commit operation.
+      if (!committerService.awaitTermination(120, TimeUnit.MINUTES)) {
+        LOG.warn(
+            "Commit operation did not complete within 120 minutes of the all files "
+                + "being rewritten. This may mean that some changes were not successfully committed to the "
+                + "table.");
+        timeout = true;
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(
+          "Cannot complete commit for rewrite, commit service interrupted", e);
+    }
+
+    if (!completedRewrites.isEmpty() && timeout) {
+      LOG.error("Attempting to cleanup uncommitted file groups");
+      completedRewrites.forEach(this::abortFileGroup);
+    }
+
+    Preconditions.checkArgument(
+        !timeout && completedRewrites.isEmpty(),
+        "Timeout occurred when waiting for commits to complete. "
+            + "{} file groups committed. {} file groups remain uncommitted. "
+            + "Retry this operation to attempt rewriting the failed groups.",
+        committedRewrites.size(),
+        completedRewrites.size());
+
+    Preconditions.checkState(
+        completedRewrites.isEmpty(),
+        "File groups offered after service was closed, " + "they were not successfully committed.");
+  }
+
+  private void commitReadyCommitGroups() {
+    Set<T> batch = null;
+    if (canCreateCommitGroup()) {
+      synchronized (completedRewrites) {
+        if (canCreateCommitGroup()) {
+          batch = Sets.newHashSetWithExpectedSize(rewritesPerCommit);
+          for (int i = 0; i < rewritesPerCommit && !completedRewrites.isEmpty(); i++) {
+            batch.add(completedRewrites.poll());
+          }
+        }
+      }
+    }
+
+    if (batch != null) {
+      String inProgressCommitToken = UUID.randomUUID().toString();
+      inProgressCommits.add(inProgressCommitToken);
+      try {
+        commitOrClean(batch);
+        committedRewrites.addAll(batch);
+      } catch (Exception e) {
+        LOG.error("Failure during rewrite commit process, partial progress enabled. Ignoring", e);
+      }
+      inProgressCommits.remove(inProgressCommitToken);
+    }
+  }
+
+  private boolean canCreateCommitGroup() {
+    // Either we have a full commit group, or we have completed writing and need to commit
+    // what is left over
+    return (completedRewrites.size() >= rewritesPerCommit)
```

Comment:
What about defining either 2 temp vars with meaningful names or two methods for individual conditions? It is a bit hard to read such complex multi-line conditions.

Reply:
Removed comment and added booleans: fullCommitGroup and writingComplete